File: src/Web/JavaScriptJS/decorator.js
class PlainMessage {
  render() {
    return "alert";
  }
}

class ComponentDecorator {
  constructor(inner) {
    this.inner = inner;
  }
}

class AuditDecorator extends ComponentDecorator {
  render() {
    return `audit(${this.inner.render()})`;
  }
}

class EncryptDecorator extends ComponentDecorator {
  render() {
    return `enc(${this.inner.render()})`;
  }
}

const component = new PlainMessage();
console.log(`base=${component.render()}`);
console.log(`audit=${new AuditDecorator(component).render()}`);
console.log(`encrypted=${new EncryptDecorator(component).render()}`);
console.log(`stacked=${new AuditDecorator(new EncryptDecorator(component)).render()}`);

File: src/Scripting/RubyRB/decorator.rb
class PlainMessage
  def render
    'alert'
  end
end

class ComponentDecorator
  def initialize(inner)
    @inner = inner
  end
end

class AuditDecorator < ComponentDecorator
  def render
    "audit(#{@inner.render})"
  end
end

class EncryptDecorator < ComponentDecorator
  def render
    "enc(#{@inner.render})"
  end
end

component = PlainMessage.new
puts "base=#{component.render}"
puts "audit=#{AuditDecorator.new(component).render}"
puts "encrypted=#{EncryptDecorator.new(component).render}"
puts "stacked=#{AuditDecorator.new(EncryptDecorator.new(component)).render}"
